Fix finding xml help files for winmds in Visual Studio.

balajikris opened the pull request:
We should pass the location of the metadata file to the
IVsXmlMemberIndexService than passing the location of the xml help file.
IVsXmlMemberIndexService locates the help file by looking at
appropriate locations such as localized directories or the root
directory where the metadata file is present. This change removes
calling our utility to locate the help file in the VS provider (the helper
is probably meant for internal test applications and such). This fixes a
bug where we were unable to find xml help files for WinMds in the fall
back case where people place help file next to winmds for cultures that
help file wasn't translated for. In such cases, the user should get the
english documentation.

**Fixes internal bug #105440**: XML doc-comment file not being picked up for WinRT components

**Tested scenarios**:
* xml documentation works in the fallback case for dlls and winmds (both Win81 and Windows UWP apps).
* xml documentation works in the localized case (help file within en folder) for dlls and winmds (both Win81 and Windows UWP apps).
* Internal test app and hosted interactive window inside of it work as expected.
* Interactive window in VS *will regress* because it does not work off of reference assemblies, it is tracked by #5013.

balajikris commented:
test vsi please

balajikris commented:
@brettfo @CyrusNajmabadi  @davkean @DustinCampbell @dpoeschl @rchande @Pilchie @jasonmalinowski @jmarolf : please review

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/MetadataReferences/VisualStudioMetadataReference.Snapshot.cs:
```diff
@@ -82,15 +82,7 @@ private bool SaveMetadataReadingException(Exception e)
             protected override DocumentationProvider CreateDocumentationProvider()
             {
-                string xmlDocumentPath;
-                if (ReferencePathUtilities.TryFindXmlDocumentationFile(FilePath, out xmlDocumentPath))
```
To avoid this mistake happening again, should we move the helper somewhere else?

davkean replied:
Yeah, can we remove the helper or move it to ETA?

davkean replied:
Funnily enough it looks like we getting lucky in the past passing the XML file to this method anyway - looking at it's implementation, it's expecting a module file path.

balajikris replied:
looks like this helper is a linked file (workspaces.desktop and scripting(portable)) and scripting uses it for [MetadataShadowCopying](https://github.com/dotnet/roslyn/blob/4b0669ac892800b041c49bdd4bb3336112b7f912/src/Scripting/Core/AssemblyLoader/MetadataShadowCopyProvider.cs#L453).

But sure, I can clean up a bit (remove code that scripting doesn't need) and move it to to scripting from workspaces.

davkean replied:
So scripting's usage is now incorrect, right?

balajikris replied:
No and Yes. It isn't using it in the context of an IVsXmlMemberIndexService. It is using it to just shadow copy metadata files along with the help files. It will be incorrect in the case where it'd miss shadow copying an xml help file for a winmd which is placed right next to it, but I'll fix that.

sounds right?

jasonmalinowski commented:
:+1:

balajikris commented:
Tagging @amcasey as he may be interested. Andrew, summary here is that you shared the locating xml help file finding utility between scripting and workspaces with d1d7303 and I'm moving it completely to scripting with 0dcaf40 since scripting is the only caller now.

jasonmalinowski commented:
@balajikris: isn't there still a caller for `MSBuildWorkspace`?

balajikris commented:
@jasonmalinowski : good question. I didn't find any references to the helper other than scripting, so i moved it.

I looked up MSBuildProjectLoader, now that you asked. It uses another XmlFileResolver :sweat: .. It's the compiler's public api..

See [MSBuildProjectLoader caller](https://github.com/dotnet/roslyn/blob/d8fbce467e757c53dc00c1862dd9c4ff6eada094/src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs#L449) and the [XmlFileResolver utility](https://github.com/dotnet/roslyn/blob/2f2629ba3e26128a15f6df9b667475a160fa5f9d/src/Compilers/Core/Portable/XmlFileResolver.cs#L46)..

Should i file a workitem for scripting to reuse this XmlFileResolver?

balajikris commented:
test mac please

balajikris commented:
Initial look at the compiler's XmlFileResolver tells me that it doesn't look in subdirectories based on culture but the internal helper does (which was possibly written to suit the IDE need and scripting started using it as well). Not sure if scripting part actually needs that functionality.

balajikris commented:
This one still needs work.. Please hold off on reviews.

Todo:
* Test Interactive and Scripting are not impacted.

balajikris commented:
@davkean @jasonmalinowski I've addressed all comments. There is a corresponding closed change as well PR 172 in the closed repo. Please take a look.

tagging @Pilchie since jason is out.

balajikris commented:
FYI, @amcasey @tmat

CyrusNajmabadi commented on src/Scripting/Core/XmlFileResolverForAssemblies.cs:
```diff
@@ -62,7 +60,7 @@ public static bool TryFindXmlDocumentationFile(string assemblyFilePath, out stri
 
             string referenceAssemblyFilePath;
-            if (!TryGetReferenceFilePath(assemblyFilePath, out referenceAssemblyFilePath))
+            if (!Roslyn.Utilities.ReferencePathUtilities.TryGetReferenceFilePath(assemblyFilePath, out referenceAssemblyFilePath))
```
Why fully qualify this?

balajikris replied:
if-def'd code. `Roslyn.Utilities` namespace is needed only where this code is active and not anywhere outside of the if-def'd region.. Also, the only occurrence, so just qualified in place.

Pilchie commented:
:+1:

balajikris commented:
test win please
